lib/compat.php

```diff
@@ -24,18 +24,90 @@
 	 * @return WP_Block_Type|false The registered block type on success, or false on failure.
 	 */
 	function register_block_type_from_metadata( $path, $args = array() ) {
-		$file = trailingslashit( $path ) . 'block.json';
-		if ( ! file_exists( $file ) ) {
+		$metadata_file = trailingslashit( $path ) . 'block.json';
+		if ( ! file_exists( $metadata_file ) ) {
 			return false;
 		}
 
-		$metadata = json_decode( file_get_contents( $file ), true );
-		if ( ! is_array( $metadata ) ) {
+		$metadata = json_decode( file_get_contents( $metadata_file ), true );
+		if ( ! is_array( $metadata ) || ! $metadata['name'] ) {
 			return false;
 		}
 
+		$block_dir           = dirname( $metadata_file );
+		$block_name          = $metadata['name'];
+		$asset_handle_prefix = str_replace( '/', '-', $block_name );
+
+		if ( isset( $metadata['editorScript'] ) ) {
+			$editor_script            = $metadata['editorScript'];
+			$editor_script_handle     = "$asset_handle_prefix-editor-script";
+			$editor_script_asset_path = realpath(
+				$block_dir . '/' . substr_replace( $editor_script, '.asset.php', -3 )
+			);
+			if ( ! file_exists( $editor_script_asset_path ) ) {
+				throw new Error(
+					"The asset file for the \"editorScript\" defined in \"$block_name\" block definition is missing."
+				);
+			}
+			$editor_script_asset  = require( $editor_script_asset_path );
+			wp_register_script(
+				$editor_script_handle,
+				plugins_url( $editor_script, $metadata_file ),
+				$editor_script_asset['dependencies'],
+				$editor_script_asset['version']
+			);
+			unset( $metadata['editorScript'] );
+			$metadata['editor_script'] = $editor_script_handle;
+		}
+
+		if ( isset( $metadata['script'] ) ) {
```

Comment:
I know that there is some code duplication here and for styles, as well. I felt resistant to extract it to shared functions because it would increase the complexity due to all the variables necessary to make it work. If you think it would read better with a helper function for scripts and another one for styles, I'm happy to refactor it.

Reply:
If anything, I'm thinking most about how we can have some automated testing for this, and wondering if we could even just extract some utility functions (e.g. for generating handle names, asset paths, etc), which would likely be easier to test than everything else dealing with filesystem and script registration.

Reply:
That's an interesting point of view. Indeed, it would be a good reason to extract this function to test some unhappy execution paths and leave the happy one to the test that covers `register_block_type_from_metadata`.

Reply:
Done in https://github.com/WordPress/gutenberg/pull/22519/commits/66e6422c3f322452f7d167d47165379e7b29ce73.